Add working bubble model for reionization

yonboyage opened the pull request:
Bubble mass function, barriers, iterations, and ionized maps.

yonboyage committed:
Add working bubble model for reionization

JulianBMunoz commented on requirements.txt:
```diff
@@ -4,4 +4,8 @@ scipy
-astropy
\ No newline at end of file
+astropy
+tqdm
```
remove

JulianBMunoz commented on tests/test_astrophysics.py:
```diff
@@ -128,6 +132,9 @@ def test_background():
 
     assert( (Coeffs.gamma_index2D >= 0.0).all()) #effective biases have to be larger than 0 in reasonable models, since galaxies live in haloes that are more clustered than average matter (in other words, SFRD grows monotonically with density)
```
what about an assert like this for the gamma_nion and gamma2 indices? Not sure if they have to always be positive, but at least that they're not nan

JulianBMunoz commented on zeus21/maps.py:
```diff
@@ -2,18 +2,22 @@
 from scipy.interpolate import interp1d
+from scipy.interpolate import InterpolatedUnivariateSpline as spline
 from pyfftw import empty_aligned as empty
+from tqdm import trange
```
remove

JulianBMunoz commented on zeus21/maps.py:
```diff
@@ -2,18 +2,22 @@
+from scipy.interpolate import InterpolatedUnivariateSpline as spline
 from pyfftw import empty_aligned as empty
+from tqdm import trange
+import time
```
remove

JulianBMunoz commented on zeus21/maps.py:
```diff
@@ -108,18 +112,129 @@ def __init__(self, T21_coefficients, Power_Spectrum, z, Lbox=600, Nbox=200, KIND
+    """
+    
+    #Measure time elapsed from start
+    start_time = time.time()
```
why?

JulianBMunoz commented on zeus21/maps.py:
```diff
@@ -108,18 +112,129 @@ def __init__(self, T21_coefficients, Power_Spectrum, z, Lbox=600, Nbox=200, KIND
+    #Measure time elapsed from start
+    start_time = time.time()
+    
+    if timer:
```
I dont mind the if timer thing, since its a optional flag shouldn it be capitalized? Also, put the time.time() inside the if

JulianBMunoz commented on zeus21/maps.py:
```diff
@@ -108,18 +112,129 @@ def __init__(self, T21_coefficients, Power_Spectrum, z, Lbox=600, Nbox=200, KIND
+
+    return ion_fields, ion_frac
+
+#look over this again
```
?

JulianBMunoz commented on zeus21/maps.py:
```diff
@@ -108,18 +112,129 @@ def __init__(self, T21_coefficients, Power_Spectrum, z, Lbox=600, Nbox=200, KIND
+    """
+    Generates a 3D map of ionized fields and ionized fraction of hydrogen.
+    
+    Uses a density threshold barrier determined from a converged bubble mass function. With default parameters, the code takes about 20 minutes on laptop to run.
```
say for all z is 20 min

JulianBMunoz commented on zeus21/maps.py:
```diff
@@ -108,18 +112,129 @@ def __init__(self, T21_coefficients, Power_Spectrum, z, Lbox=600, Nbox=200, KIND
+    spherize: bool
+        Whether to flag spheres around ionized cells (True) instead of only central pixel flagging (False). Default is False. Central pixel flagging is generally more consistent with the bubble mass function than spherizing.
+
+    Outputs
```
maybe output the z too since they ay be different than input as it chooses closest

JulianBMunoz commented on zeus21/reionization.py:
```diff
@@ -0,0 +1,279 @@
+        self.sigma = CoeffStructure.sigmaofRtab
+
+        self.Hz = cosmology.Hubinvyr(CosmoParams, self.zlist)
+        #self.Hzconv = (u.km/u.s/u.Mpc).to(1/u.yr) #multiply this times Hubble parameter to convert from km/s/Mpc to 1/yr
```
remove

JulianBMunoz commented on zeus21/reionization.py:
```diff
@@ -0,0 +1,279 @@
+        self.trec = self.trec0/(1+self.zlist)**3/constants.yrTos #years
+        self.niondot_avg = CoeffStructure.niondot_avg_II
+
+        #self.xHI = CoeffStructure.xHI_avg ----> need to figure out what's going on with this
```
?

JulianBMunoz commented on zeus21/reionization.py:
```diff
@@ -0,0 +1,279 @@
+        self.ion_frac = np.fmin(1, [self.calc_Q(CosmoParams, i) for i in np.arange(len(self.zlist))])
+        self.ion_frac_initial = np.copy(self.ion_frac)
+
+        zr_mesh = np.meshgrid(np.arange(len(self.Rs)), np.arange(len(self.zlist)))
```
explain what this is in a comment

JulianBMunoz commented on zeus21/reionization.py:
```diff
@@ -0,0 +1,279 @@
+            nion_values = self.nion_delta_r_int(ds_array, ir)  #Shape (nd, nz)
+            nrec_values = self.nrec(ds_array, ion_frac)             #Shape (nd, nz)
+        
+            total_values = np.log10(nion_values / (1 + nrec_values))   #taking difference in logspace to find zero-crossing 
```
explain why 1+nrec!

JulianBMunoz commented on zeus21/reionization.py:
```diff
@@ -0,0 +1,279 @@
+
+
+    #normalizing the nion/sfrd model
+    def nion_normalization(self, z, R):
```
explain what is being normalized, its so that <Nion(delta)> equals the average

JulianBMunoz commented on zeus21/reionization.py:
```diff
@@ -0,0 +1,279 @@
+        integrand_base = denom * ion_frac_rev 
+    
+        # Broadcast over d_array
+        integrand = integrand_base[np.newaxis, :] * (1 + d_array[:, np.newaxis])
```
growth

JulianBMunoz commented on zeus21/reionization.py:
```diff
@@ -0,0 +1,279 @@
+
+        self.ZMAX_REION = ZMAX_REION #max redshift up to which we calculate reionization observables
+        self.zlist = CoeffStructure.zintegral
+        self.Rs = CoeffStructure.Rtabsmoo
```
question: is this a good idea? we may want to have a different R array for bubbles, no? For instance, we don't care about R>100 Mpc or so, whereas Rtabsmoo goes much farther. Why don't we set a RMAX_Bubbles on UserParams and have self.Rs just be the Rtabsmoo smaller than that?

yonboyage committed:
Fixed map function

yonboyage committed:
Added gamma2 for SFRD_II and _III

yonboyage committed:
Changes to BMF, setup.py, z21utilities, and added
ionized maps class

EmilieThelie committed:
Fixed minor issues with reionization_maps.

EmilieThelie committed:
Added zreion/treion maps computation.

EmilieThelie committed:
Fixes for zreion/treion maps.

EmilieThelie committed:
Fixes for zreion/treion maps.

EmilieThelie committed:
Fixes for zreion/treion maps.

EmilieThelie committed:
Include partial ionization in reionization_maps.

EmilieThelie committed:
Fixes for partial ionization.

EmilieThelie committed:
Fixes for partial ionization.

EmilieThelie committed:
Fixes for partial ionization.

EmilieThelie committed:
Fixes for partial ionization.

EmilieThelie committed:
Fix for partial ionization.

yonboyage committed:
Changes to reionization.py:

-Functions take values instead of indexes
-Added interpolators
-Added Rmin and ZMAX

yonboyage committed:
Changes to maps.py

- Removed enforce_BMF_scale, allowing user to pick any box size and resolution, but forcing smoothing Rs.
- Allow input Rs but comes with a warning.
- Using linear barrier by default (uses new functions from reionization.py).
- Updated partial ionizations, and added their massweightings.
- Small fixes.

yonboyage committed:
Add ionized maps to __init__.py

yonboyage committed:
Merge branch 'main' into BMF_and_ionized_maps

yonboyage committed:
Update maps.py

yonboyage committed:
change np.trapz to np.trapezoid

yonboyage committed:
Fixed some test calls

yonboyage committed:
More trapz to trapezoid, and also uncommented import empty

yonboyage committed:
Added required import

yonboyage committed:
pbox import

yonboyage committed:
Added reionization map tests

JulianBMunoz commented on tests/test_astrophysics.py:
```diff
@@ -15,6 +15,7 @@
 
 from zeus21.sfrd import *
 from zeus21.correlations import *
+from zeus21.reionization import *
```
you should write more tests, basically for all functions/classes

JulianBMunoz commented on zeus21/cosmology.py:
```diff
@@ -115,6 +115,38 @@ def runclass(CosmologyIn):
     
     return ClassCosmo
 
+def time_at_redshift(ClassyCosmo,z):
```
This seems pretty inefficient, it's reading the ts and zs every time it's called and then interpolating. Why not do the reading once in CosmoParams (since it does the same for chi comoving distance vs z) and then this function just evaluats the interpolator with CosmoParams.zlistage and CosmoParams.tlistages saved?

JulianBMunoz commented on zeus21/maps.py:
```diff
@@ -108,18 +112,394 @@ def __init__(self, T21_coefficients, Power_Spectrum, z, Lbox=600, Nbox=200, KIND
+        Number of cells on a side. Default is 300 cells.
+    seed: int
+        Sets the predetermined generation of maps. Default is 1234.
+    r_precision: float
```
is this in Mpc? or relative to something? clarify

JulianBMunoz commented on zeus21/maps.py:
```diff
@@ -108,18 +112,394 @@ def __init__(self, T21_coefficients, Power_Spectrum, z, Lbox=600, Nbox=200, KIND
+            if self.r[0] < smallest_r:
+                print(f'WARNING: Your input radii are too small for the pixel size. The code will still run now.\nIn the future, for best performance and physical accuracy on this boxlength and ncells, the smallest smoothing radius should be no less than R=L/N * (4pi/3)^(-1/3), or approximately {smallest_r:.2f} cMpc.')
+        
+        self.seed = seed
```
what if the user wants a random seed but not to input it? maybe set the default seed as None, will that break things below?

JulianBMunoz commented on zeus21/maps.py:
```diff
@@ -108,18 +112,394 @@ def __init__(self, T21_coefficients, Power_Spectrum, z, Lbox=600, Nbox=200, KIND
+            z21_utilities.print_timer(start_time, text_before="    done in ")
+        return density_field
+
+    def generate_density_allz(self, CosmoParams):
```
are you generating a delta box at each z? why? you can just use the same box and multiply by growth on the fly no?

JulianBMunoz commented on zeus21/z21_utilities.py:
```diff
@@ -0,0 +1,52 @@
+
+    return realmap
+
+def tophat_smooth(rr, ks, dk):
```
need to explain what it takes and what it returns. It took me a while to understand dk = field in k space and it returns field in x space smoothed

JulianBMunoz replied:
isnt this defined in maps.py? maybe centralize to utilities

JulianBMunoz commented on zeus21/z21_utilities.py:
```diff
@@ -0,0 +1,52 @@
+    x = ks * rr + 1e-5
+    win_k = 3/(x**3) * (np.sin(x) - x*np.cos(x))
+    deltakfilt = dk * win_k
+    return np.real(np.fft.ifftn(deltakfilt))
```
have you tested whether np.abs and np.real are different? just as an extra check do it once

JulianBMunoz commented on zeus21/maps.py:
```diff
@@ -108,18 +112,394 @@ def __init__(self, T21_coefficients, Power_Spectrum, z, Lbox=600, Nbox=200, KIND
+            curr_z_idx = self._z_idx[i]
+            ion_field = self.ionize(CosmoParams, CoeffStructure, curr_z_idx)
+            ion_field_allz[i] = ion_field
+            ion_frac[i] = np.sum(ion_field)/(self.ncells**3)
```
have you tried putting mass weighted ion_frac here? may be fast

JulianBMunoz commented on zeus21/maps.py:
```diff
@@ -108,18 +112,394 @@ def __init__(self, T21_coefficients, Power_Spectrum, z, Lbox=600, Nbox=200, KIND
+        iterator = trange(len(self.z)) if self.PRINT_TIMER else range(len(self.z))
+        
+        for i in iterator:
+            curr_z_idx = self._z_idx[i]
```
what's happening here? why is _z_idz different from i ?

JulianBMunoz commented on zeus21/maps.py:
```diff
@@ -108,18 +112,394 @@ def __init__(self, T21_coefficients, Power_Spectrum, z, Lbox=600, Nbox=200, KIND
+        Dg0 = CosmoParams.growthint(zlist[0])
+        Dg = CosmoParams.growthint(zlist[curr_z_idx])
+        if not self.SPHERIZE:
+            ion_field = np.any(self.density_smoothed_allr > (Dg0/Dg)*self.barrier[curr_z_idx, self._r_idx][:, None, None, None], axis=0)
```
have you tried np.sum(x // 1 ) instead of np.any to see if it's any faster? im curious

JulianBMunoz commented on zeus21/maps.py:
```diff
@@ -108,18 +112,394 @@ def __init__(self, T21_coefficients, Power_Spectrum, z, Lbox=600, Nbox=200, KIND
+        for i in iterator:
+            nion_spl = spline(sample_d, BMF.nion_delta_r_int(CosmoParams, sample_d, self.z, r)[:, i])
+            nrec_spl = spline(sample_d, BMF.nrec(CosmoParams, sample_d, BMF.ion_frac, self.z)[:, i])
+            partial_ion_spl = spline(sample_d, nion_spl(sample_d)/(1+nrec_spl(sample_d)))
```
do you need to re-spline everything?

JulianBMunoz replied:
also, do you need 51 densities? I suspect if you spline in (nion/(1+nrec)) you won't need as much to be precise, especially if you do log[] vs delta. Not sure if that is a bottleneck or not

JulianBMunoz commented on zeus21/maps.py:
```diff
@@ -108,18 +112,394 @@ def __init__(self, T21_coefficients, Power_Spectrum, z, Lbox=600, Nbox=200, KIND
+            
+        return self.ion_frac_partial, self.ion_field_partial_allz
+
+    def compute_partial_massweighted(self, CosmoParams, BMF, r=None):
```
is this used when doing partial + massweighted? I'm confused

JulianBMunoz replied:
because if doing partial and MW should be clipped differently

JulianBMunoz commented on zeus21/reionization.py:
```diff
@@ -0,0 +1,351 @@
+class BMF:
+    """
+    Computes the bubble mass function (BMF). 
+
```
This needs some documentation :)

JulianBMunoz commented on zeus21/reionization.py:
```diff
@@ -0,0 +1,351 @@
+
+        zr_mesh = np.meshgrid(np.arange(len(self.Rs)), np.arange(len(self.zlist)))
+        self.nion_norm = self.nion_normalization(zr_mesh[1], zr_mesh[0])
+        self.nion_norm_int = RegularGridInterpolator(self.zr, self.nion_norm, bounds_error = False, fill_value = np.nan)
```
have you checked if it's worth interpolating the log instead of nion?

JulianBMunoz commented on zeus21/reionization.py:
```diff
@@ -0,0 +1,351 @@
+        self.BMF = np.array([self.VRdn_dR(z, self.Rs_BMF) for z in self.zlist]) #initial bubble mass function
+        self.BMF_initial = np.copy(self.BMF)
+        self.ion_frac = np.nan_to_num([np.trapezoid(self.BMF[i], np.log(self.Rs_BMF)) for i in range(len(self.zlist))]) #ion_frac by integrating the BMF
+        self.ion_frac[self.barrier[:, -1]<=0] = 1
```
does this matter? have you tested what happens if you don't do it?

JulianBMunoz commented on zeus21/reionization.py:
```diff
@@ -0,0 +1,351 @@
+            The resultant density threshold array. First dimension is each redshift, second dimension is each radius scale.
+        """
+        barrier = np.zeros((len(z), len(R)))
+        ds_array = np.linspace(-1, 5, 101)
```
you need 100 densities?

JulianBMunoz commented on zeus21/reionization.py:
```diff
@@ -0,0 +1,351 @@
+        z = z[zarg]
+        ion_frac = ion_frac[zarg]
+        
+        for ir in range(len(R)):
```
why dont Ir, RR in  enumerate(R) ?

JulianBMunoz commented on zeus21/reionization.py:
```diff
@@ -0,0 +1,351 @@
+                idx = np.where(sign_change)[0]
+                if idx.size > 0:
+                    #Linear interpolation to find zero crossings
+                    x0 = ds_array[idx]
```
if idx is where it changes sign shouldnt we go 1 below instead up above?

JulianBMunoz commented on zeus21/reionization.py:
```diff
@@ -0,0 +1,351 @@
+                #Find zero crossings
+                sign_change = np.diff(np.sign(y_values))
+                idx = np.where(sign_change)[0]
+                if idx.size > 0:
```
what if idx is exactly at the edge? ds_array[idx+1] will give an error

JulianBMunoz commented on zeus21/reionization.py:
```diff
@@ -0,0 +1,351 @@
+        return (barriermax - barriermin)/(sigmax**2 - sigmin**2)
+        
+    def B_0(self, z):
+        sigmin = self.sigmaR_int(z, self.Rs[self.R_linear_sigma_fit_idx-1])
```
why R_linear_sigma_fit_idx - 1 here?

JulianBMunoz commented on zeus21/reionization.py:
```diff
@@ -0,0 +1,351 @@
+
+    #interpolators in z and R used in reionization.py
+    
+    def interpR(self, z, R, func):
```
question, does this belong in utilities? I'm not sure.

JulianBMunoz replied:
also given that its logR rename to logR so its clear

JulianBMunoz commented on zeus21/z21_utilities.py:
```diff
@@ -0,0 +1,52 @@
+    deltakfilt = dk * win_k
+    return np.real(np.fft.ifftn(deltakfilt))
+
+def find_nearest_idx(array, values):
```
same

JulianBMunoz commented on zeus21/z21_utilities.py:
```diff
@@ -0,0 +1,52 @@
+        idx.append((np.abs(array - values[i])).argmin())
+    return np.unique(idx)
+
+def print_timer(start_time, text_before="", text_after=""):
```
and same

JulianBMunoz commented on zeus21/z21_utilities.py:
```diff
@@ -0,0 +1,52 @@
+    secs = int(elapsed_time - mins*60)
+    print(f"{text_before}{mins}min {secs}s{text_after}")
+
+def v2r(v):
```
this could also be in utilities, for reionization.py

JulianBMunoz commented on zeus21/sfrd.py:
```diff
@@ -225,8 +227,9 @@ def __init__(self, User_Parameters, Cosmo_Parameters, ClassCosmo, Astro_Paramete
-        SFRD_II_dR = np.trapz(integrand_II, HMF_interpolator.logtabMh, axis = 2)
+        # Compute SFRD and niondot quantities
+        SFRD_II_dR = np.trapezoid(integrand_II, HMF_interpolator.logtabMh, axis = 2)
+        niondot_II_dR = np.trapezoid(integrand_II*fesctab_II[None, None, :, None], HMF_interpolator.logtabMh, axis = 2)
```
isnt there an ionizing efficiency here? generically it could depend on Mh so let's bring it here if it was elsewhere

JulianBMunoz commented on zeus21/sfrd.py:
```diff
@@ -235,20 +238,53 @@ def __init__(self, User_Parameters, Cosmo_Parameters, ClassCosmo, Astro_Paramete
-        self.gamma_II_index2D = np.log(SFRD_II_dR[:,:,-1]/SFRD_II_dR[:,:,0]) / (deltaArray[:,:,0,-1] - deltaArray[:,:,0,0])
+        midpoint = deltaArray.shape[-1]//2 #midpoint of deltaArray at delta = 0
+
+        self.gamma_II_index2D = np.log(SFRD_II_dR[:,:,midpoint+1]/SFRD_II_dR[:,:,midpoint-1]) / (deltaArray[:,:,0,midpoint+1] - deltaArray[:,:,0,midpoint-1])
```
since we repeat this (and the 2nd derivative) a lot, is it worth functionalizing? take array, deltarray, midpoint, and return everything

JulianBMunoz commented on zeus21/sfrd.py:
```diff
@@ -585,8 +629,8 @@ def __init__(self, User_Parameters, Cosmo_Parameters, ClassCosmo, Astro_Paramete
         
-        self.niondot_avg_II = Astro_Parameters.N_ion_perbaryon_II/cosmology.rho_baryon(Cosmo_Parameters,0.) * np.trapz(integrand_II_table * fesctab_II, HMF_interpolator.logtabMh, axis = 1)
-        self.niondot_avg_III = Astro_Parameters.N_ion_perbaryon_II/cosmology.rho_baryon(Cosmo_Parameters,0.) * np.trapz(integrand_III_table * fesctab_III, HMF_interpolator.logtabMh, axis = 1)
+        self.niondot_avg_II = Astro_Parameters.N_ion_perbaryon_II/cosmology.rho_baryon(Cosmo_Parameters,0.) * np.trapezoid(integrand_II_table * fesctab_II, HMF_interpolator.logtabMh, axis = 1)
```
let's discuss this, I'd prefer to parametrize this through xiion directly in nion. That said, I'm confused why this is repeated w.r.t. above

JulianBMunoz reviewed (commented):
Great job Yonny and Emilie! This looks like pro coding. There are some physics questions in the comments (plus the slack). There are also some suggestions for docs, and also for optimization. Let's discuss it, but you can start working on it in the meantime

yonboyage committed:
Fixd single-z map computation and small R partials

Fixed single-z map computation by precomputing partial ionizations as a function of density in reionization.py

Allowed extrapolation of small R values in partial ionizations. This lets the user make more highly resolved maps than before.

yonboyage committed:
Added print option for successful run

yonboyage committed:
Extend density power in CoevalMaps

yonboyage committed:
Altered partial ionizations

maps.py: density array used for partial ionizations and massweighting now the density field smoothed by the pixel scale.

reionization.py: minor tweaks for self-consistency

yonboyage committed:
Updated Modeling of Maps and Bubbles

Maps:
- Added mass variance correction to account for lack of ergodicity in smaller simulations
- Fixed partial ionization computation accounting for this sigma correction.

Bubbles:
- Changed all interpolators to allow for linear extrapolation
- Added new moving pivot for linear barrier to account for the typical bubble size changing at different redshifts. Now the typical bubbles are the ones best approximated in the linear barrier.
- Now computing the ionized fraction from the analytic solution to the integral of the BMF. This accounts for contribution from bubbles outside the set radius range rather than integrating numerically.
- Increased tolerance for wiggle between iterations of the BMF.

yonboyage committed:
Small fix

Forgot to add input to function

EmilieThelie committed:
Added clumping as arg to AstroParams, inputs for
densities in reionization_maps, and
delete_class_attributes to z21_utilities.
Removed density_allz from partials in
reionization_maps.
Created astro_variations class in
reionization_maps.
Changed CosmoParams.Rsmmin to 0.05.

EmilieThelie committed:
Fixes for the type for the barrier input in
reionization_maps.

EmilieThelie committed:
Improve astro_variations
+ increase kmax in test_correlations
+ remove classy.class from cosmo_wrapper

EmilieThelie committed:
Small fix in astro_variations.

EmilieThelie committed:
Small fix in astro_variations.

yonboyage committed:
Added broadcasting to most code

- Broadcasting used to speed up computation by 3x, almost everywhere
- Several BMF functions now accept arrays as well as scalars
- Added moving pivot for barrier fits, so that now the BMF best represents the bubbles around the peak, rather than a static fit at one bubble size that has increasing errors farther from the pivot; i.e. the most common bubbles are always well-modeled by the new moving pivot.
- Added a few interpolators, including for peak R of BMF
